fix: only flag object assign in accumulators, if it is allocating a new object for each iter

emilyinure opened the pull request:
… object on every assignment

Closes #6675

<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve?-->
Accumulating Object.Assign isn't inherently bad as not all calls create new objects which result in the O(n*n) memory usage.
<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->
#6675 
#5277

<!-- If you create a user-facing change, please write a changeset: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#writing-a-changeset (your changeset is often a good starting point for this summary as well) -->

## Test Plan

<!-- What demonstrates that your implementation is correct? -->
Testing shows that only when we are creating a new object every iteration, and assigning to it from the accumulator, it gets flagged

<!-- If you're submitting a new rule or action (or an option for them), the documentation is part of the code. Make sure rules and actions have example usages, and that all options are documented. -->

<!-- For other features, please submit a documentation PR to the `next` branch of our website: https://github.com/biomejs/website/. Link the PR here once it's ready. -->

emilyinure committed:
fix: only flag object assign in accumulators, if it is creating a new object on every assignment

#6675

emilyinure committed:
fix: addictional handling for object assign inside of accumulators using new keyword to allocate target objects

emilyinure committed:
fix: Satisfy clippy on accumulating spread

emilyinure committed:
noAccumulatingSpread changeset

ematipico commented:
@emilyinure any updates?

emilyinure committed:
Updated examples for docs on no-accumulating-spread

emilyinure committed:
Fix parameter name to match reduce parameter

ematipico commented on .changeset/wet-jeans-repair.md:
```diff
@@ -0,0 +1,5 @@
+"@biomejs/biome": patch
+---
+
+Fix #6675: Only flag noAccumulatingSpread on Object.assign when a new object is being allocated on each iteration. Ex: foo.reduce((acc, bar) => Object.assign({}, acc, bar), {})
```
The contribution guide explains how to write the changeset (the PR template has a link to that paragraph):
- sentences should start with the past tense
- link the issue 
- examples should be inside a code block

Refers to other changesets as examples

emilyinure committed:
Merge branch 'main' into fix-accumulating-object-creation

emilyinure committed:
Update noAccumulatingSpread changeset

Updated tense usage at the beginning of changeset, linked back to issue.
Moved examples to the inside of codeblocks
Explains which cases aren't caught anymore and which are.

emilyinure committed:
Update changeset for noAccumulatingSpread

Remove example that contains syntax error
Update grammer

dyc3 commented on .changeset/wet-jeans-repair.md:
````diff
@@ -2,4 +2,22 @@
+
+The following cases which **do** make new allocations will continue to warn:
+
+```js,expect_diagnostic
````
nit: We don't check diagnostics in changesets
```suggestion
```js
```

dyc3 commented on .changeset/wet-jeans-repair.md:
````diff
@@ -2,4 +2,22 @@
+```js,expect_diagnostic
+foo.reduce((acc, bar) => Object.assign({}, acc, bar), {})
+```
+```js,expect_diagnostic
````
nit: same here
```suggestion
```js
```

emilyinure committed:
Apply suggestion from @dyc3

Co-authored-by: Carson McManus <dyc3@users.noreply.github.com>
